### PR TITLE
(Re)create endpoints in a stable order

### PR DESCRIPTION
**Pull request by akerouanton**
Related to:

- #20179
- #45906 
- moby/libnetwork#2093
- https://github.com/moby/moby/issues/20067

Fixes:

- https://github.com/moby/moby/issues/48868

**- What I did**

* 1st commit: Add a new `Priority` field to `EndpointSettings` ;
* 2nd commit:
  * Use this new field to make sure the calls to `daemon.connectToNetwork` are done in descending order ;
  * Also, use it to make sure the endpoints join the sandbox in the same order (that's what guarantees interface names are stable over restarts) ;
* 3rd commit: On `NetworkConnect`, the `Priority` value is automatically computed based on the lowest `Priority` minus one ;

The choice to use descending over ascending order has been made to match what the Compose spec defines.

This change brings interesting advantages for end users:

* Since the interface name/order is now stable, users can bind a specific interface without needing an external script to resolve what interface should be used ;
* The DNS resolver will always iterate on a container's networks in the exact same order, hence DNS resolution for a given name present on multiple networks a container is connected to will always resolve to the same thing ;

**- How to verify it**

* CI should be green.

**- Description for the changelog**

Interfaces are always created in the same order and their name is stable over restarts.

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: api/server/router/network/network_routes.go [hunk 1]
```diff
@@ -258,7 +258,7 @@ func (n *networkRouter) postNetworkConnect(ctx context.Context, w http.ResponseW
 	// The reason is that, In case of attachable network in swarm scope, the actual local network
 	// may not be available at the time. At the same time, inside daemon `ConnectContainerToNetwork`
 	// does the ambiguity check anyway. Therefore, passing the name to daemon would be enough.
-	return n.backend.ConnectContainerToNetwork(connect.Container, vars["id"], connect.EndpointConfig)
+	return n.backend.ConnectContainerToNetwork(connect.Container, vars["id"], connect.EndpointConfig, true)
```
Comment (akerouanton) [inline api/server/router/network/network_routes.go:261]: I'm not sure whether this should be set to `false` for API < 1.44 :thinking:

Comment (thaJeztah) [reply api/server/router/network/network_routes.go:261]: If the current order is randomised by go (thus "undefined"), then applying things in the correct order would make it equally "undefined" (just 100% more predictable), so I think it's fine to make it independent of the API version.

The Priority _FIELD_ should not be looked at for older API versions (when creating).

File: daemon/container_operations.go [hunk 2]
```diff
@@ -1052,13 +1068,20 @@ func errRemovalContainer(containerID string) error {
 }
 
 // ConnectToNetwork connects a container to a network
-func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName string, endpointConfig *networktypes.EndpointSettings) error {
+func (daemon *Daemon) ConnectToNetwork(container *container.Container, idOrName string, endpointConfig *networktypes.EndpointSettings, autoPriority bool) error {
```
Comment (thaJeztah) [inline daemon/container_operations.go:1071]: Looks like `autoPriority` is unused?

Comment (akerouanton) [reply daemon/container_operations.go:1071]: Oh, good catch!

File: api/server/router/network/network_routes.go (hunk 1 above)
Comment (akerouanton) [reply api/server/router/network/network_routes.go:261]: > The Priority FIELD should not be looked at for older API versions (when creating).

Mh, I'm assuming `Priority` is 0 for older API versions but that might be only for Go client :thinking: Maybe I should force it to 0 for older versions.

**Comment by akerouanton** [comment]
So, one of the failing test is checking if the default gateway and "anonymous" port mappings change whenever the container connects to a network which comes before other networks the container is connected to in the lexicographic order:

```
$ docker network create testnet1
$ docker network create aaaa
$ docker run -d --name test -p 9000:90 -p 70 busybox top

$ docker port test
70/tcp -> 0.0.0.0:32774
70/tcp -> [::]:32774
90/tcp -> 0.0.0.0:9000
90/tcp -> [::]:9000
$ docker exec test ip -o route show      
default via 172.18.0.1 dev eth0 
172.18.0.0/16 dev eth0 scope link  src 172.18.0.2 

$ docker network connect testnet1 test
$ docker port test                       
70/tcp -> 0.0.0.0:32774
70/tcp -> [::]:32774
90/tcp -> 0.0.0.0:9000
90/tcp -> [::]:9000
$ docker exec test ip -o route show      
default via 172.18.0.1 dev eth0 
172.18.0.0/16 dev eth0 scope link  src 172.18.0.2 
172.19.0.0/16 dev eth3 scope link  src 172.19.0.2 

# Here we go:
$ docker network connect aaaa test 
$ docker port test                 
70/tcp -> 0.0.0.0:32775
70/tcp -> [::]:32775
90/tcp -> 0.0.0.0:9000
90/tcp -> [::]:9000
$ docker exec test ip -o route show
default via 172.21.0.1 dev eth4 
172.18.0.0/16 dev eth0 scope link  src 172.18.0.2 
172.19.0.0/16 dev eth3 scope link  src 172.19.0.2 
172.21.0.0/16 dev eth4 scope link  src 172.21.0.2 
```

And I'm pretty sure some people rely on this to set their default gateway. I need to check if that's still documented (it was at some point) and what to do next.

**Comment by mbrodala** [comment]
Is there still work in progress here? Or something missing?

**Comment by thaJeztah** [comment]
superseded by https://github.com/moby/moby/pull/48936

**Pull request closed without merging**